modules/fakemeta/pdata_gc.cpp

```diff
@@ -0,0 +1,537 @@
+// vim: set ts=4 sw=4 tw=99 noet:
+//
+// AMX Mod X, based on AMX Mod by Aleksander Naszko ("OLO").
+// Copyright (C) The AMX Mod X Development Team.
+//
+// This software is licensed under the GNU General Public License, version 3 or higher.
+// Additional exceptions apply. For full license details, see LICENSE.txt or visit:
+//     https://alliedmods.net/amxmodx-license
+
+//
+// Fakemeta Module
+//
+
+#include "fakemeta_amxx.h"
+#include <am-algorithm.h>
+
+enum class BaseFieldType
+{
+	None,
+	Integer,
+	Float,
+	Vector,
+	Entity,
+	String,
+};
+
+static const char *BaseFieldTypeName[] =
+{
+	"none",
+	"integer",
+	"float",
+	"vector",
+	"entity",
+	"string",
+};
+
+static BaseFieldType GetBaseDataType(TypeDescription &data)
+{
+	switch (data.fieldType)
+	{
+		case FieldType::FIELD_INTEGER:
+		case FieldType::FIELD_STRINGINT:
+		case FieldType::FIELD_SHORT:
+		case FieldType::FIELD_CHARACTER:
+		case FieldType::FIELD_CLASS:
+		case FieldType::FIELD_STRUCTURE:
+		case FieldType::FIELD_POINTER:
+		case FieldType::FIELD_FUNCTION:
+		case FieldType::FIELD_BOOLEAN:
+		{
+			return BaseFieldType::Integer;
+		}
+		case FieldType::FIELD_FLOAT:
+		{
+			return BaseFieldType::Float;
+		}
+		case FieldType::FIELD_VECTOR:
+		{
+			return BaseFieldType::Vector;
+		}
+		case FieldType::FIELD_CLASSPTR:
+		case FieldType::FIELD_ENTVARS:
+		case FieldType::FIELD_EDICT:
+		case FieldType::FIELD_EHANDLE:
+		{
+			return BaseFieldType::Entity;
+		}
+		case FieldType::FIELD_STRINGPTR:
+		case FieldType::FIELD_STRING:
+		{
+			return BaseFieldType::String;
+		}
+	}
+
+	return BaseFieldType::None;
+}
+
+#define GET_TYPE_DESCRIPTION(position, data, baseType)                                     \
+	int classLength, memberLength;                                                         \
+	const char *className  = MF_GetAmxString(amx, params[position + 1], 0, &classLength);  \
+	const char *memberName = MF_GetAmxString(amx, params[position + 2], 1, &memberLength); \
```

Comment:
Why is this offset by 1/2 instead of 0/1? It looks like you are retrieving the same parameter twice in code:

```C++
int entity = params[1];
CHECK_ENTITY(entity);

TypeDescription data;
GET_TYPE_DESCRIPTION(1, data, BaseFieldType::Integer);
```